Documentation/Porting.md

```diff
@@ -0,0 +1,249 @@
+
+```diff
+--- a/Sources/_TestingInternals/include/Includes.h
++++ b/Sources/_TestingInternals/include/Includes.h
+
++#if defined(macintosh)
++#include <DateTimeUtils.h>
++#endif
+```
+
+We intentionally don't import platform-specific C standard library modules
+(`Darwin`, `Glibc`, `WinSDK`, etc.) in Swift because they often include overlay
+code written in Swift and adding those modules as dependencies would make it
+more difficult to test that Swift code using Swift Testing. 
+
+### Changes in Swift
+
+Once the header is included, we can call `GetDateTime()` from `Clock.swift`:
+
+```diff
+--- a/Sources/Testing/Events/Clock.swift
++++ b/Sources/Testing/Events/Clock.swift
+
+ fileprivate(set) var wall: TimeValue = {
+ #if !SWT_NO_TIMESPEC
+   // ...
++#elseif os(Classic)
++  var seconds = CUnsignedLong(0)
++  GetDateTime(&seconds)
++  seconds -= 2_082_844_800 // seconds between epochs
++  return TimeValue((seconds: Int64(seconds), attoseconds: 0))
+ #else
+ #warning("Platform-specific implementation missing: UTC time unavailable (no timespec)")
+ #endif
+ }
+```
+
+## Runtime test discovery
+
+When porting to a new platform, you may need to provide a new implementation for
+`enumerateTypeMetadataSections()` in `Discovery.cpp`. Test discovery is
+dependent on Swift metadata discovery which is an inherently platform-specific
+operation.
+
+_Most_ platforms will be able to reuse the implementation used by Linux and
+Windows that calls an internal Swift runtime function to enumerate available
+metadata. If you are porting Swift Testing to Classic, this function won't be
+available, so you'll need to write a custom implementation instead. Assuming
+that the Swift compiler emits section information into the resource fork on
+Classic, you could use the [Resource Manager](https://developer.apple.com/library/archive/documentation/mac/pdf/MoreMacintoshToolbox.pdf)
+to load that information:
+
+```diff
+--- a/Sources/_TestingInternals/Discovery.cpp
++++ b/Sources/_TestingInternals/Discovery.cpp
+
+ // ...
++#elif defined(macintosh)
++template <typename SectionEnumerator>
++static void enumerateTypeMetadataSections(const SectionEnumerator& body) {
++  ResFileRefNum refNum;
++  if (noErr == GetTopResourceFile(&refNum)) {
++    ResFileRefNum oldRefNum = refNum;
++    do {
++      UseResFile(refNum);
++      Handle handle = Get1NamedResource('swft', "\p__swift5_types");
++      if (handle && *handle) {
++        size_t size = GetHandleSize(handle);
++        body(*handle, size);
++      }
++    } while (noErr == GetNextResourceFile(refNum, &refNum));
++    UseResFile(oldRefNum);
++  }
++}
+ #else
+ #warning Platform-specific implementation missing: Runtime test discovery unavailable
+ template <typename SectionEnumerator>
+ static void enumerateTypeMetadataSections(const SectionEnumerator& body) {}
+ #endif
+```
+
+## C++ stub implementations
+
+Some symbols defined in C and C++ headers, especially "complex" macros, cannot
```

Comment:
I've sometimes seen these referred to as "non-trivial" macros

Reply:
Six of one, half-dozen of the other?